src/vm/checker/check_db.rs

```diff
@@ -1,145 +1,110 @@
+    pub fn execute <F, T, E> (&mut self, f: F) -> Result<T,E> where F: FnOnce(&mut Self) -> Result<T,E>, {
+        self.begin();
+        let result = f(self)
+            .or_else(|e| {
+                self.roll_back();
+                Err(e)
+            })?;
+        self.commit();
+        Ok(result)
     }
 
-    pub fn inner_open(filename: &str) -> AnalysisDatabaseConnection {
-        let conn = Connection::open(filename)
-            .expect("Failure to open analysis db.");
-        AnalysisDatabaseConnection {
-            conn: conn
-        }
+    pub fn begin(&mut self) {
+        self.store.nest();
     }
 
-    pub fn begin_save_point(&mut self) -> AnalysisDatabase<'_> {
-        AnalysisDatabase::from_savepoint(self.conn.savepoint()
-                                         .expect(SQL_FAIL_MESSAGE))
+    pub fn commit(&mut self) {
+        self.store.commit();
     }
-}
 
-impl <'a> AnalysisDatabase <'a> {
-
-    pub fn from_savepoint(sp: Savepoint<'a>) -> AnalysisDatabase<'a> {
-        AnalysisDatabase {
-            savepoint: sp }
+    pub fn roll_back(&mut self) {
+        self.store.rollback();
     }
 
-    pub fn roll_back(&mut self) {
-        self.savepoint.rollback()
-            .expect(SQL_FAIL_MESSAGE);
+    fn put(&mut self, key: &str, value: &ClaritySerializable) {
+        self.store.put(&key, &value.serialize());
     }
 
-    pub fn commit(self) {
-        self.savepoint.commit()
-            .expect(SQL_FAIL_MESSAGE);
+    fn get <T> (&mut self, key: &str) -> Option<T> where T: ClarityDeserializable<T> {
+        self.store.get(&key)
+            .map(|x| T::deserialize(&x))
     }
 
-    pub fn execute<P>(&mut self, sql: &str, params: P) -> usize
-    where
-        P: IntoIterator,
-        P::Item: ToSql {
-        self.savepoint.execute(sql, params)
-            .expect(SQL_FAIL_MESSAGE)
+    fn make_key(contract_name: &str) -> String {
+        format!("analysis::{}", contract_name)
     }
 
-    fn query_row<T, P, F>(&self, sql: &str, params: P, f: F) -> Option<T>
-    where
-        P: IntoIterator,
-        P::Item: ToSql,
-        F: FnOnce(&Row) -> T {
-        self.savepoint.query_row(sql, params, f)
-            .optional()
-            .expect(SQL_FAIL_MESSAGE)
+    fn load_contract(&mut self, contract_name: &str) -> Option<ContractAnalysis> {
+        let key = AnalysisDatabase::make_key(contract_name);
+        self.get(&key)
     }
 
-    fn load_contract(&self, contract_name: &str) -> Option<ContractAnalysis> {
-        let result: Option<String> = self.query_row(
-            "SELECT analysis FROM type_analysis_table WHERE contract_name = ?",
-            &[contract_name],
-            |row| row.get(0));
-        match result {
-            Some(contract) => Some(ContractAnalysis::deserialize(&contract)),
-            None => None
+    pub fn insert_contract(&mut self, contract_name: &str, contract: &ContractAnalysis) -> CheckResult<()> {
```

Comment:
Do we wan to check that the contract name is well-formed here, or is that handled for-sure by the caller?

Reply:
I think when we address #1078, this method signature will change to require a fully qualified contract principal as input, that way it'll be enforced by the type system. I'll add a note in that issue.